docs/source/asr/datasets.rst

```diff
@@ -633,6 +633,135 @@ Some other Lhotse related arguments we support:
 
 The full and always up-to-date list of supported options can be found in ``LhotseDataLoadingConfig`` class.
 
+Extended multi-dataset and multi-modal configuration format
+~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
+
+Combining a large number of datasets and defining weights for them can be tricky.
+We offer an extended configuration format that allows you to explicitly define datasets,
+dataset groups, and their weights either inline in the experiment configuration,
+or as a path to a separate YAML file.
+
+In addition to the features above, this format introduces a special ``tags`` dict-like field.
+The keys and values in ``tags`` are automatically attached to every sampled example, which
+is very useful when combining multiple datasets with different properties.
+The dataset class which converts these examples to tensors can partition the mini-batch and apply
+different processing to each group.
+For example, you may want to construct different prompts for the model using metadata in ``tags``.
+
+Example 1. Combine two datasets with equal weights and attach custom metadata in ``tags`` to each cut:
+
+.. code-block:: yaml
+
+    input_cfg:
+      - type: nemo_tarred
+        manifest_filepath: /path/to/manifest__OP_0..512_CL_.json
+        tarred_audio_filepath: /path/to/tarred_audio/audio__OP_0..512_CL_.tar
+        weight: 0.4
+        tags:
+          lang: en
+          pnc: no
+      - type: nemo_tarred
+        manifest_filepath: /path/to/other/manifest__OP_0..512_CL_.json
+        tarred_audio_filepath: /path/to/other/tarred_audio/audio__OP_0..512_CL_.tar
+        weight: 0.6
+        tags:
+          lang: pl
+          pnc: yes
+
+Example 2. Combine multiple (4) datasets, corresponding to different tasks (ASR, AST).
+Each task gets its own group and its own weight.
+Then within each task, each dataset get its own within-group weight as well.
+The final weight is the product of outer and inner weight:
+
+.. code-block:: yaml
+
+    input_cfg:
+      - type: group
+        weight: 0.7
+        tags:
+          task: asr
+        input_cfg:
+          - type: nemo_tarred
+            manifest_filepath: /path/to/asr1/manifest__OP_0..512_CL_.json
+            tarred_audio_filepath: /path/to/tarred_audio/asr1/audio__OP_0..512_CL_.tar
+            weight: 0.6
+            tags:
```

Comment:
Yes, it should be the configs and/or the docs.

Reply:
Let's keep it in the docs primarily detailing all allowed components. Configs can have and should have them, but we need to take extraordinary care to keep them updated then

Reply:
@pzelasko should we  add something like `validate_tags` to model class dataloaders to keep track of `tags` supported by each model class?

Reply:
@titu1994 +1

@krishnacpuvvada Yes but not in the model class. Since model classes are abstract, it's the same model class supporting Canary-1B and some future model that might require a different set of tags. This validation can be done inside of `prompt_format_fn` which (at least for multi-task AED models) is the main component using these tags.